File: lua/test/submit_entity_test.lua
-- Submit entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("jokeapi_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("SubmitEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:Submit(nil)
    assert.is_not_nil(ent)
  end)

  it("should run basic flow", function()
    local setup = submit_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"create"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "submit." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set JOKEAPI_TEST_SUBMIT_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- CREATE
    local submit_ref01_ent = client:Submit(nil)
    local submit_ref01_data = helpers.to_map(vs.getprop(
      vs.getpath(setup.data, "new.submit"), "submit_ref01"))

    local submit_ref01_data_result, err = submit_ref01_ent:create(submit_ref01_data, nil)
    assert.is_nil(err)
    submit_ref01_data = helpers.to_map(submit_ref01_data_result)
    assert.is_not_nil(submit_ref01_data)

  end)
end)

function submit_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/submit/SubmitTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read submit test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "submit01", "submit02", "submit03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("JOKEAPI_TEST_SUBMIT_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["JOKEAPI_TEST_SUBMIT_ENTID"] = idmap,
    ["JOKEAPI_TEST_LIVE"] = "FALSE",
    ["JOKEAPI_TEST_EXPLAIN"] = "FALSE",
    ["JOKEAPI_APIKEY"] = "NONE",
  })

  local idmap_resolved = helpers.to_map(
    env["JOKEAPI_TEST_SUBMIT_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["JOKEAPI_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
        apikey = env["JOKEAPI_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["JOKEAPI_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["JOKEAPI_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end

File: ts/test/entity/submit/SubmitEntity.test.ts


import Path from 'node:path'
import * as Fs from 'node:fs'

import { test, describe, afterEach } from 'node:test'
import assert from 'node:assert'
import { createLiveTransport } from '../../live-runner'
import { runLiveEntity } from '../../live-entity'


import { JokeapiSDK, BaseFeature, stdutil } from '../../..'

import {
  envOverride,
  liveClientOptions,
  liveDelay,
  loadEnvLocal,
  makeCtrl,
  makeMatch,
  makeReqdata,
  makeStepData,
  makeValid,
  maybeSkipControl,
} from '../../utility'


// AFTER the imports on purpose: TypeScript hoists `import` above any
// statement in the emitted CommonJS, so a loader placed above them would
// run only after every imported module had already been evaluated - and
// anything reading process.env at module scope would miss these values.
loadEnvLocal(__dirname + '/../../../.env.local')


describe('SubmitEntity', async () => {

  // Per-test live pacing. Delay is read from sdk-test-control.json's
  // `test.live.delayMs`; only sleeps when JOKEAPI_TEST_LIVE=TRUE.
  afterEach(liveDelay('JOKEAPI_TEST_LIVE'))

  test('instance', async () => {
    const testsdk = JokeapiSDK.test()
    const ent = testsdk.Submit()
    assert(null != ent)
  })


  test('basic', async (t) => {

    const live = 'TRUE' === process.env.JOKEAPI_TEST_LIVE
    for (const op of ['create']) {
      if (!live && maybeSkipControl(t, 'entityOp', 'submit.' + op, live)) return
    }

    
    const setup = basicSetup()
    if (setup.live) {
      return runLiveEntity(setup, {"active":true,"alias":{"field":{}},"fields":[{"active":true,"name":"category","req":true,"type":"`$STRING`","index$":0},{"active":true,"name":"delivery","req":false,"short":"The delivery/punchline (for twopart jokes only)","type":"`$STRING`","index$":1},{"active":true,"name":"error","req":false,"type":"`$BOOLEAN`","index$":2},{"active":true,"name":"flags","req":true,"type":"`$OBJECT`","index$":3},{"active":true,"name":"formatVersion","req":true,"type":"`$INTEGER`","index$":4},{"active":true,"name":"joke","req":false,"short":"The joke content (for single-type jokes only)","type":"`$STRING`","index$":5},{"active":true,"name":"lang","req":true,"short":"Language code (ISO 639-1)","type":"`$STRING`","index$":6},{"active":true,"name":"message","req":false,"type":"`$STRING`","index$":7},{"active":true,"name":"setup","req":false,"short":"The setup part (for twopart jokes only)","type":"`$STRING`","index$":8},{"active":true,"name":"timestamp","req":false,"short":"13-character UNIX timestamp","type":"`$INTEGER`","index$":9},{"active":true,"name":"type","req":true,"type":"`$STRING`","index$":10}],"name":"submit","op":{"create":{"input":"data","name":"create","points":[{"active":true,"args":{"query":[{"active":true,"kind":"query","name":"dry_run","orig":"dry_run","reqd":false,"type":"`$BOOLEAN`","index$":0},{"active":true,"example":"json","kind":"query","name":"format","orig":"format","reqd":false,"type":"`$STRING`","index$":1}]},"contract":{"id":"POST /submit","json":"{\"operationId\":\"submitJoke\",\"parameters\":[{\"description\":\"If present, validates the joke without actually saving it to the API.\",\"in\":\"query\",\"name\":\"dry-run\",\"required\":false,\"schema\":{\"type\":\"boolean\"}},{\"description\":\"Response format. Defaults to JSON if not specified.\",\"in\":\"query\",\"name\":\"format\",\"required\":false,\"schema\":{\"default\":\"json\",\"enum\":[\"json\",\"xml\",\"yaml\",\"txt\"],\"type\":\"string\"}}],\"protocol\":\"http\",\"requestBody\":{\"content\":{\"application/json\":{\"schema\":{\"properties\":{\"category\":{\"enum\":[\"Misc\",\"Programming\",\"Dark\",\"Pun\",\"Spooky\",\"Christmas\"],\"example\":\"Programming\",\"type\":\"string\"},\"delivery\":{\"description\":\"The delivery/punchline (for twopart jokes only)\",\"type\":\"string\"},\"flags\":{\"properties\":{\"explicit\":{\"description\":\"Contains explicit language\",\"example\":false,\"type\":\"boolean\"},\"nsfw\":{\"description\":\"Not safe for work\",\"example\":false,\"type\":\"boolean\"},\"political\":{\"description\":\"Contains political content\",\"example\":false,\"type\":\"boolean\"},\"racist\":{\"description\":\"Contains racist content\",\"example\":false,\"type\":\"boolean\"},\"religious\":{\"description\":\"Contains religious content\",\"example\":false,\"type\":\"boolean\"},\"sexist\":{\"description\":\"Contains sexist content\",\"example\":false,\"type\":\"boolean\"}},\"required\":[\"nsfw\",\"religious\",\"political\",\"racist\",\"sexist\",\"explicit\"],\"type\":\"object\"},\"formatVersion\":{\"example\":3,\"type\":\"integer\"},\"joke\":{\"description\":\"The joke content (for single-type jokes only)\",\"type\":\"string\"},\"lang\":{\"description\":\"Language code (ISO 639-1)\",\"example\":\"en\",\"type\":\"string\"},\"setup\":{\"description\":\"The setup part (for twopart jokes only)\",\"type\":\"string\"},\"type\":{\"enum\":[\"single\",\"twopart\"],\"example\":\"single\",\"type\":\"string\"}},\"required\":[\"formatVersion\",\"category\",\"type\",\"flags\",\"lang\"],\"type\":\"object\"}}},\"required\":true},\"responses\":{\"200\":{\"content\":{\"application/json\":{\"schema\":{\"properties\":{\"error\":{\"example\":false,\"type\":\"boolean\"},\"message\":{\"example\":\"Joke submitted successfully\",\"type\":\"string\"},\"timestamp\":{\"description\":\"13-character UNIX timestamp\",\"example\":1234567890123,\"type\":\"integer\"}},\"type\":\"object\"}}},\"description\":\"Joke submitted successfully or dry-run validation successful\"},\"400\":{\"content\":{\"application/json\":{\"schema\":{\"properties\":{\"additionalInfo\":{\"type\":\"string\"},\"causedBy\":{\"items\":{\"type\":\"string\"},\"type\":\"array\"},\"code\":{\"example\":400,\"type\":\"integer\"},\"error\":{\"example\":true,\"type\":\"boolean\"},\"internalError\":{\"example\":false,\"type\":\"boolean\"},\"message\":{\"example\":\"Invalid parameter value\",\"type\":\"string\"},\"timestamp\":{\"description\":\"13-character UNIX timestamp\",\"type\":\"integer\"}},\"required\":[\"error\",\"code\",\"message\"],\"type\":\"object\"}}},\"description\":\"Bad request - invalid joke submission\"},\"429\":{\"content\":{\"application/json\":{\"schema\":{\"properties\":{\"additionalInfo\":{\"type\":\"string\"},\"causedBy\":{\"items\":{\"type\":\"string\"},\"type\":\"array\"},\"code\":{\"example\":400,\"type\":\"integer\"},\"error\":{\"example\":true,\"type\":\"boolean\"},\"internalError\":{\"example\":false,\"type\":\"boolean\"},\"message\":{\"example\":\"Invalid parameter value\",\"type\":\"string\"},\"timestamp\":{\"description\":\"13-character UNIX timestamp\",\"type\":\"integer\"}},\"required\":[\"error\",\"code\",\"message\"],\"type\":\"object\"}}},\"description\":\"Rate limit exceeded - maximum 5 requests per minute for submissions\"}},\"securitySource\":\"unspecified\"}","source":"openapi3","version":1},"kind":"http","method":"POST","orig":"/submit","segments":[{"lit":"submit"}],"select":{"exist":["dry_run","format"]},"transform":{"req":"`reqdata`","res":"`body`"},"index$":0}],"key$":"create"}},"relations":{"ancestors":[]},"key$":"submit","name__orig":"submit","Name":"Submit","name_":"submit","name-":"submit","NAME":"SUBMIT","index$":2}, {"active":true,"entity":"submit","key$":"BasicSubmitFlow","kind":"basic","name":"BasicSubmitFlow","param":{},"step":[{"active":true,"data":{},"input":{"ref":"submit_ref01"},"match":{},"op":"create","spec":[],"valid":[],"index$":0}]}, 'Submit')
    }
    const client = setup.client
    const struct = setup.struct

    const isempty = struct.isempty
    const select = struct.select


    // CREATE
    const submit_ref01_ent = client.Submit()
    let submit_ref01_data = setup.data.new.submit['submit_ref01']

    submit_ref01_data = (await submit_ref01_ent.create(submit_ref01_data)).data()
    assert(null != submit_ref01_data)


  })
})



function basicSetup(extra?: any) {
  // TODO: fix test def options
  const options: any = {} // null

  // TODO: needs test utility to resolve path
  const entityDataFile =
    Path.resolve(__dirname, 
      '../../../../.sdk/test/entity/submit/SubmitTestData.json')

  // TODO: file ready util needed?
  const entityDataSource = Fs.readFileSync(entityDataFile).toString('utf8')

  // TODO: need a xlang JSON parse utility in voxgig/struct with better error msgs
  const entityData = JSON.parse(entityDataSource)

  options.entity = entityData.existing

  let client = JokeapiSDK.test(options, extra)
  const struct = client.utility().struct
  const merge = struct.merge
  const transform = struct.transform

  let idmap = transform(
    ['submit01','submit02','submit03'],
    {
      '`$PACK`': ['', {
        '`$KEY`': '`$COPY`',
        '`$VAL`': ['`$FORMAT`', 'upper', '`$COPY`']
      }]
    })

  const env = envOverride({
    'JOKEAPI_TEST_SUBMIT_ENTID': idmap,
    'JOKEAPI_TEST_LIVE': 'FALSE',
    'JOKEAPI_TEST_EXPLAIN': 'FALSE',
  })

  idmap = env['JOKEAPI_TEST_SUBMIT_ENTID']

  const live = 'TRUE' === env.JOKEAPI_TEST_LIVE

  const transport = createLiveTransport()
  if (live) {
    const rawIds = process.env['JOKEAPI_TEST_SUBMIT_ENTID']
    idmap = rawIds && rawIds.trim() ? JSON.parse(rawIds) : {}
    if (!idmap || Array.isArray(idmap) || typeof idmap !== 'object') {
      throw new Error('Live ENTID must be a JSON object')
    }
    client = new JokeapiSDK(merge([
      // FIRST, so the generated fields below win: sdk-test-control.json's
      // test.client.options adds to the live client, it does not redirect it.
      liveClientOptions(),
      {
      },
      // 'extra || {}', not a bare 'extra': struct.merge returns UNDEFINED when the
      // last entry is undefined, and basicSetup is normally called with no
      // argument at all - so a bare 'extra' silently discarded the apikey
      // and server values above and handed the SDK undefined. Harmless
      // while there was nothing in that object; not harmless now.
      extra || {},
      { system: { fetch: transport.fetch } }
    ]))
  }

  const setup = {
    idmap,
    env,
    options,
    client,
    struct,
    data: entityData,
    explain: 'TRUE' === env.JOKEAPI_TEST_EXPLAIN,
    live,
    transport,
    now: Date.now(),
  }

  return setup
}
  
